File: NeTracker.py
from bcc import BPF
from time import sleep
import socket
import struct

#this where I load eBPF program
bpf_program = """
#include <uapi/linux/ptrace.h>
#include <net/sock.h>
#include <bcc/proto.h>

int kprobe__tcp_v4_connect(struct pt_regs *ctx, struct sock *sk)
{
    u32 pid = bpf_get_current_pid_tgid();
    u32 saddr = sk->__sk_common.skc_rcv_saddr;
    u32 daddr = sk->__sk_common.skc_daddr;
    u16 sport = sk->__sk_common.skc_num;
    u16 dport = bpf_ntohs(sk->__sk_common.skc_dport);
    bpf_trace_printk("TCPv4 Connect: PID=%d SADDR=%d DADDR=%d SPORT=%d DPORT=%d\\n", 
                     pid, saddr, daddr, sport, dport);
    return 0;
}
int trace_close(struct pt_regs *ctx, struct sock *sk)
{
    u32 pid = bpf_get_current_pid_tgid();
    u32 saddr = sk->__sk_common.skc_rcv_saddr;
    u32 daddr = sk->__sk_common.skc_daddr;
    u16 sport = sk->__sk_common.skc_num;
    u16 dport = bpf_ntohs(sk->__sk_common.skc_dport);

    bpf_trace_printk("TCP Close: PID=%d SADDR=%d DADDR=%d SPORT=%d DPORT=%d\\n", 
                     pid, saddr, daddr, sport, dport);
    return 0;
}
"""

b = BPF(text=bpf_program)

#this is where I attach the eBPF program to the kernel
b.attach_kprobe(event="tcp_v4_connect", fn_name="trace_connect")
b.attach_kprove(event="tcp_close", fn_name="trace_close")

#headers
print("%-6s %-12s %-16s %-16s %-6s %-6s" % ("PID", "COMM", "SADDR", "DADDR", "SPORT", "DPORT"))

#this to read and process data from eBPF map
while True:
    try:
        sleep(1)
        with open("/sys/kernel/debug/tracing/trace_pipe", "r") as f:
            for line in f:
                print(line.strip())
    except KeyboardInterrupt:
        print("\nStopping")
        break

File: NeTracker.c
#include <asm/ptrace.h>
#include <asm/tcp.h>
#include <net/sock.h>
#include <bpf/bpf_helpers.h>

//this where I define the data structure that will be used to store the connection information
struct data_t {
    u32 pid;
    u32 saddr;
    u32 daddr;
    u16 sport;
    u16 dport;
    u64 start_time;
};

//defining the map that will be used to store the connection information
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 1024);
    __type(key, u32);
    __type(value, struct data_t);
} connections SEC(".maps");

//this is the function that will handle the connection establishment
//it will be called when a connection is established
SEC("kprobe/tcp_v4_connect")
int trace_connect(struct pt_regs *ctx, struct sock *sk) {
    u32 pid = bpf_get_current_pid_tgid();
    struct data_t data = {};
    
    data.pid = pid;
    data.saddr = sk->__sk_common.skc_rcv_saddr;
    data.daddr = sk->__sk_common.skc_daddr;
    data.sport = sk->__sk_common.skc_num;
    data.dport = bpf_ntohs(sk->__sk_common.skc_dport);
    data.start_time = bpf_ktime_get_ns();

    bpf_map_update_elem(&connections, &pid, &data, BPF_ANY);

    //print connection info to the trace pipe
    bpf_trace_printk("Connect: PID=%d SADDR=%d DADDR=%d SPORT=%d DPORT=%d\\n", 
                    data.pid, data.saddr, data.daddr, data.sport, data.dport);

    return 0;
    }

//this is the hook into function that handles connection close
SEC("kprobe/tcp_close")
int trace_close(struct pt_regs *ctx, struct sock *sk) {
    u32 pid = bpf_get_current_pid_tgid();
    struct data_t *data = bpf_map_lookup_elem(&connections, &pid);

    if (data) {
        u64 duration = bpf_ktime_get_ns() - data->start_time;
        
        //this is to print connection duration in kernel trace
        bpf_trace_printk("Close: PID=%d SADDR=%d DADDR=%d SPORT=%d DPORT=%d Duration=%llu ns\\n", 
                         data->pid, data->saddr, data->daddr, data->sport, data->dport, duration);

        bpf_map_delete_elem(&connections, &pid);
    }

    return 0;
}

char LICENSE[] SEC("license") = "GPL";